### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by 35niavlys** [vaadin/vaadin-menu-bar#89]
Performance is very bad compared to that of vaadin 7 in a big DOM. Are improvements planned on this subject? Impossible to go into production for my part

**Comment by tomivirkki** [comment]
Hi, are you referring to performance of Vaadin genereally or specifically to `vaadin-menu-bar` component? What kind of performance issues have you faced? Some specific browser?

**Comment by 35niavlys** [comment]
Hi, I refer to vaadin-menu-bar. My context is a menubar with around 600 items in page with a lot of svg charts. Chrome is more or less ok but slower than vaadin 7. Firefox 60.8.0esr is really slow. I often see the menu open at top-left position (or last position) then the good position. So I cannot navigate easily in the tree

**Comment by tomivirkki** [comment]
Sounds like an unusual use case. Created a menu bar with 600 items here and it indeed is sluggish https://shimmering-gallium.glitch.me/ (understandably)

Maybe some other component would better fit your use case better since menu bar isn't designed for that many items and hasn't built-in virtual scrolling. Maybe a menu-bar with an iron-list as the only sub-item. Or a context-menu with an iron-list...

**Comment by 35niavlys** [comment]
Items are not all in one submenu, it's a hierachy. I have 15 items max per submenu but each item can have subitems...

**Comment by tomivirkki** [comment]
Can you modify the example to showcase the issue?

**Comment by 35niavlys** [comment]
Your exemple is slow too but less than on my app:
![ff](https://user-images.githubusercontent.com/881049/63846395-9e956380-c983-11e9-87f6-2acff1ac58ac.gif)

On my fat app:
- vaadin7:
![ff_vaadin7](https://user-images.githubusercontent.com/881049/63846454-c1277c80-c983-11e9-8951-f8bb3b812af8.gif)
- vaadin14
![ff_vaadin14](https://user-images.githubusercontent.com/881049/63846462-c5539a00-c983-11e9-8810-62a5304ee90a.gif)
:

**Comment by tomivirkki** [comment]
Updated the example to use hierarchical items. It's working fine on FF also. https://shimmering-gallium.glitch.me/ Wonder if the sluggishness results from something else. Do you use Flow / Java to build the menubar structure?

**Comment by 35niavlys** [comment]
Yes I use flow. On little page it works fine but the difference with vaadin 7 on the same page is huge. I think the main problem is when the mouse move from first item to last item fastly, all events are processed. In fact we maybe need to process only the last event and not all the queue of events. I don't know if I'm understandable

I generate this page. Works better than my own app but...
```public class MenuBarRoute extends Div {

	public MenuBarRoute() {
		add(generateMenuBar());
		
		HorizontalLayout h = new HorizontalLayout();
		add(h);
		
		for (int i = 0 ; i++ < 10;) {
			VerticalLayout v = new VerticalLayout();
			h.add(v);
			for (int j = 0 ; j++ < 100;) {
				v.add(new Button("button-"+i, e -> Notification.show(e.toString())));
			}
		}
	}

	private static MenuBar generateMenuBar() {
		MenuBar menu = new MenuBar();
		for (int i = 0 ; i++ < 10;)
			generateSubMenu(menu.addItem("item-" + i), 2);
		return menu;
	}

	private static void generateSubMenu(MenuItem item, int level) {
		level--;
		for (int i = 0 ; i++ < 10;) {
			MenuItem subItem = item.getSubMenu().addItem(item.getText() + "-" + i);
			if(level > 0)
				generateSubMenu(subItem, level);
		}
	}
}
